src/ompl/geometric/planners/rrt/src/RRTstar.cpp

```diff
@@ -573,73 +652,322 @@ void ompl::geometric::RRTstar::getPlannerData(base::PlannerData &data) const
+                }
+                else
+                {
+                    // No, so we will be pruning this motion:
+                    leavesToPrune.push_back(*mIter);
+                }
+            }
+
+            // Pop the iterator, std::list::erase returns the next iterator
+            mIter = motionQueue.erase(mIter);
+        }
+
+       // We now have a list of Motions to definitely remove, and a list of Motions to recheck
+       // Iteratively check the two lists until there is nothing to to remove
+        while (leavesToPrune.empty() == false)
         {
-            pruneScratchSpace_.newTree.push_back(candidate);
-            pruneScratchSpace_.candidates.insert(pruneScratchSpace_.candidates.end(),
-                candidate->children.begin(), candidate->children.end());
+            // First, remove the Motions
+            // Let us reuse our iterator
+            mIter = leavesToPrune.begin();
+            while (mIter != leavesToPrune.end())
+            {
+                // Remove the leaf from its parent
+                removeFromParent(*mIter);
+
+                // Erase the actual motion
+                // First free the state
+                si_->freeState((*mIter)->state);
+
+                // then delete the pointer
+                delete *mIter;
+
+                // And finally remove it from the list, erase returns the next iterator
+                mIter = leavesToPrune.erase(mIter);
+
+                // Update our counter
+                ++numPruned;
+            }
+
+            // Now, we need to go through the list of chain vertices and see if any are now leaves
+            // Again, let us reuse our iterator
+            mIter = chainsToRecheck.begin();
+            while (mIter != chainsToRecheck.end())
+            {
+                // Is the Motion a leaf?
+                if ((*mIter)->children.empty() == true)
+                {
+                    // It is, add to the removal queue
+                    leavesToPrune.push_back(*mIter);
+
+                    // Remove from this queue, getting the next
+                    mIter = chainsToRecheck.erase(mIter);
+                }
+                else
+                {
+                    // Is isn't ,skip to the next
+                    ++mIter;
+                }
+            }
+        }
+
+        // All done pruning.
+        // Update the cost at which we've pruned:
+        prunedCost_ = pruneTreeCost;
+
+        // And if we're using informed sampling, the measure to which we've pruned
+        if (useInformedSampling_)
+        {
+            prunedInfMeasure_ = infSampler_->getInformedMeasure(prunedCost_);
         }
     }
 
-    // To create the new nn takes one order of magnitude in time more than just checking how many
-    // states would be pruned. Therefore, only prune if it removes a significant amount of states.
-    if ((double)pruneScratchSpace_.newTree.size() / tree_size < pruneStatesThreshold_)
```

Comment:
@jvgomez, please correct me if I'm wrong, but my understanding is this is a condition to decide whether it is worth the effort to prune or not. It is calculated from the number of vertices that will be pruned, and as a result requires evaluating the entire tree to make a prune decision on each vertex.

The proposed change replaces this condition with one that is based on the change in cost since the last time the graph was pruned.

The advantage to a cost-based method is that it doesn't require any calculation (just a comparison of `Cost`) before deciding to prune. The disadvantage is that it doesn't know how many vertices will actually be pruned from the tree.